### PR TITLE
Added cmdlets and tests or the .NET SDK version 4.0.1

### PR DESCRIPTION
**Pull request by prakharsharma10**
## Description

The changes in this PR include
1. Added PS models for 'delivery policy' which has been added as a property on a Azure CDN endpoint
2. Added a cmdlet validating probe URL
3. Added cmdlets for enabling and disabling custom HTTPS
4. Added cmdlet for getting supported optimization types for a CDN profile
5. Added tests for all the new cmdlets

Link for .Net SDK https://github.com/Azure/azure-sdk-for-net/tree/psSdkJson6/src/SDKs/Cdn 

Cmdlet review here: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/23

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#making-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by maddieclayton** [comment]
@prakharsharma10 Have you filled out a cmdlet review for these changes? If not, please do so here: https://github.com/Azure/azure-powershell-cmdlet-review-pr

**Comment by cormacpayne** [comment]
@MiYanni feel free to unassign yourself once the cmdlet design review has completed

**Comment by MiYanni** [comment]
@prakharsharma10 @cormacpayne Design review added: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/23#issuecomment-372459073

**Comment by prakharsharma10** [comment]
@cormacpayne @MiYanni, Have updated the PR and the issue with the details of the changes.

**Comment by cormacpayne** [comment]
@prakharsharma10 there are compilation errors in the build -- would you mind taking a look?

**Comment by prakharsharma10** [comment]
@cormacpayne The compilation errors are because we are using [pattern matching](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7#pattern-matching) which was introduced in C# 7. Is this not supported yet?

**Comment by prakharsharma10** [comment]
@cormacpayne  I have fixed the issues with the markdown files. The compilation is failing because of StaticAnalysis but there are no csv files attached. Is that blocking the PR from being reviewed?

**Comment by markcowl** [comment]
@prakharsharma10 Please fix merge conflicts

**Comment by prakharsharma10** [comment]
@cormacpayne Please suggest any changes, if required.

File: src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/CustomDomainTests.ps1 [hunk 1]
```diff
@@ -19,15 +19,17 @@ Gets and removes custom domain with running endpoint.
 function Test-CustomDomainGetRemoveWithRunningEndpoint
 {
     # Hard-coding host and endpoint names due to requirement for DNS CNAME
-    $endpointName = "sdktest-c83c1e8f-343e-4ce8-873b-f6e5ddcdc53f"
-    $hostName = "sdktest-716d4572-627f-4dfe-8128-1df163647ae2.azureedge-test.net"
+	#$endpointName = "sdktest-c83c1e8f-343e-4ce8-873b-f6e5ddcdc53f"
+    #$hostName = "sdktest-716d4572-627f-4dfe-8128-1df163647ae2.azureedge-test.net"
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/CustomDomainTests.ps1]: @prakharsharma10 please remove these commented-out lines

File: src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/CustomDomainTests.ps1 [hunk 2]
```diff
@@ -59,22 +61,73 @@ function Test-CustomDomainGetRemoveWithRunningEndpoint
...
 
+<#
+.SYNOPSIS
+Enables custom domain with running endpoint.
+#>
+function Test-CustomDomainEnableDisableWithRunningEndpoint
+{
+    # Hard-coding host and endpoint names due to requirement for DNS CNAME
+    #$endpointName = "sdktest-c83c1e8f-343e-4ce8-873b-f6e5ddcdc53f"
+    #$hostName = "sdktest-716d4572-627f-4dfe-8128-1df163647ae2.azureedge-test.net"
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/CustomDomainTests.ps1]: @prakharsharma10 please remove these commented-out lines

File: src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/CustomDomainTests.ps1 [hunk 3]
```diff
@@ -59,22 +61,73 @@ function Test-CustomDomainGetRemoveWithRunningEndpoint
...
 .SYNOPSIS
 Gets and removes custom domain with stopped endpoint
 #>
 function Test-CustomDomainGetRemoveWithStoppedEndpoint
 {
     # Hard-coding host and endpoint names due to requirement for DNS CNAME
-    $endpointName = "sdktest-cbc4e6fa-da15-4f37-9511-6b7df122c1de" 
-    $hostName = "sdktest-34a59412-9044-4166-b055-d777e111e810.azureedge-test.net"  
+    #$endpointName = "sdktest-cbc4e6fa-da15-4f37-9511-6b7df122c1de" 
+    #$hostName = "sdktest-34a59412-9044-4166-b055-d777e111e810.azureedge-test.net"
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/CustomDomainTests.ps1]: @prakharsharma10 please remove these commented-out lines

File: src/ResourceManager/Cdn/Commands.Cdn/Commands.Cdn.csproj [hunk 4]
```diff
@@ -45,21 +45,26 @@
     <Prefer32Bit>false</Prefer32Bit>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Microsoft.Azure.Management.Cdn">
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Cdn.3.0.0-preview\lib\net45\Microsoft.Azure.Management.Cdn.dll</HintPath>
-      <Private>True</Private>
+    <Reference Include="Microsoft.Azure.Management.Cdn, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Cdn.4.0.1-preview\lib\net452\Microsoft.Azure.Management.Cdn.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.Rest.ClientRuntime, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL" />
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/Commands.Cdn.csproj]: @prakharsharma10 please remove the references added in this `.csproj` file (`Microsoft.Rest.ClientRuntime.*` and `Newtonsoft.Json`) since they are referenced in a common targets file

File: src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/DisableAzureRmCdnCustomDomain.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        public string ProfileName { get; set; }
+
+        [Parameter(Mandatory = true, HelpMessage = "The resource group of the Azure CDN profile.", ParameterSetName = FieldsParameterSet)]
+        [ResourceGroupCompleter()]
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, HelpMessage = "The custom domain object.", ParameterSetName = ObjectParameterSet)]
+        [ValidateNotNull]
+        public PSCustomDomain CdnCustomDomain { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/DisableAzureRmCdnCustomDomain.cs]: @prakharsharma10 this should be renamed to `InputObject` -- you can also add an alias to `CdnCustomDomain`

File: src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/EnableAzureRmCdnCustomDomain.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        public string ProfileName { get; set; }
+
+        [Parameter(Mandatory = true, HelpMessage = "The resource group of the Azure CDN profile.", ParameterSetName = FieldsParameterSet)]
+        [ResourceGroupCompleter()]
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, HelpMessage = "The custom domain object.", ParameterSetName = ObjectParameterSet)]
+        [ValidateNotNull]
+        public PSCustomDomain CdnCustomDomain { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/EnableAzureRmCdnCustomDomain.cs]: @prakharsharma10 same comment about this being called `InputObject`

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/ConfirmAzureRmCdnEndpointProbeUrl.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+using Microsoft.Azure.Commands.Cdn.Common;
+using Microsoft.Azure.Commands.Cdn.Helpers;
+using Microsoft.Azure.Commands.Cdn.Models.Endpoint;
+using Microsoft.Azure.Commands.Cdn.Properties;
+using Microsoft.Azure.Management.Cdn;
+using Microsoft.Azure.Management.Cdn.Models;
+
+namespace Microsoft.Azure.Commands.Cdn.Endpoint
+{
+    [Cmdlet(VerbsLifecycle.Confirm, "AzureRmCdnEndpointProbeURL"), OutputType(typeof(PSValidateProbeOutput))]
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/ConfirmAzureRmCdnEndpointProbeUrl.cs]: @prakharsharma10 what does this cmdlet do? If we are checking the existence of a resource or object, then the verb for this cmdlet should be `Test` instead of `Confirm`

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/GetAzureRmCdnEndpointResourceUsage.cs [hunk 8]
```diff
@@ -52,7 +52,7 @@ public override void ExecuteCmdlet()
                 ProfileName = CdnEndpoint.ProfileName;
                 EndpointName = CdnEndpoint.Name;
             }
-
+
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/GetAzureRmCdnEndpointResourceUsage.cs]: @prakharsharma10 please revert the change made in this file

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/NewAzureRmCdnEndpoint.cs [hunk 9]
```diff
@@ -90,9 +90,15 @@ public class NewAzureRmCdnEndpoint : AzureCdnCmdletBase
...
         public string OptimizationType { get; set; }
 
+        [Parameter(Mandatory = false, HelpMessage = "Specifies the probe path for Dynamic Site Acceleration")]
+        public string ProbePath { get; set; }
+
         [Parameter(Mandatory = false, HelpMessage = "The list of geo filters that applies to this endpoint.")]
         public PSGeoFilter[] GeoFilters { get; set; }
 
+        [Parameter(Mandatory = false, HelpMessage = "The delivery policy for this endpoint.")]
+        public PSDeliveryPolicy DeliveryPolicy { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/NewAzureRmCdnEndpoint.cs]: @prakharsharma10 how do users create this object?

File: src/ResourceManager/Cdn/Commands.Cdn/Profile/GetAzureRmCdnProfileSupportedOptimizationType.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+using Microsoft.Azure.Commands.Cdn.Properties;
+using Microsoft.Azure.Management.Cdn;
+using System.Linq;
+using Microsoft.Azure.Commands.Cdn.Helpers;
+using Microsoft.Azure.Commands.Cdn.Models;
+using Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters;
+
+namespace Microsoft.Azure.Commands.Cdn.Profile
+{
+    [Cmdlet(VerbsCommon.Get, "AzureRmCdnProfileSupportedOptimizationType", DefaultParameterSetName = FieldsParameterSet), OutputType(typeof(List<PSOptimizationType>))]
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/Profile/GetAzureRmCdnProfileSupportedOptimizationType.cs]: @prakharsharma10 rather than having this separate cmdlet to get additional information about a profile, we should see if we can add a `SwitchParameter` to the existing `Get-AzureRmCdnProfile` cmdlet that the user will pass to signal they want the cmdlet to make the extra API call to get information about the supported optimization types, and that can be found in an additional property of the `PSProfile` object that is returned by that cmdlet. Let me know what you think

File: src/ResourceManager/Cdn/Commands.Cdn/help/Confirm-AzureRmCdnEndpointProbeURL.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+## SYNTAX
+
+```
+Confirm-AzureRmCdnEndpointProbeURL -ProbeUrl <String> [-DefaultProfile <IAzureContextContainer>]
+ [<CommonParameters>]
+```
+
+## DESCRIPTION
+{{Fill in the Description}}
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/help/Confirm-AzureRmCdnEndpointProbeURL.md]: @prakharsharma10 please add a description for this cmdlet

File: src/ResourceManager/Cdn/Commands.Cdn/help/Disable-AzureRmCdnCustomDomain.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+```
+
+### ByObjectParameterSet
+```
+Disable-AzureRmCdnCustomDomain -CdnCustomDomain <PSCustomDomain> [-PassThru]
+ [-DefaultProfile <IAzureContextContainer>] [<CommonParameters>]
+```
+
+## DESCRIPTION
+{{Fill in the Description}}
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/help/Disable-AzureRmCdnCustomDomain.md]: @prakharsharma10 please add a description for this cmdlet

File: src/ResourceManager/Cdn/Commands.Cdn/help/Enable-AzureRmCdnCustomDomain.md [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+```
+
+### ByObjectParameterSet
+```
+Enable-AzureRmCdnCustomDomain -CdnCustomDomain <PSCustomDomain> [-PassThru]
+ [-DefaultProfile <IAzureContextContainer>] [<CommonParameters>]
+```
+
+## DESCRIPTION
+{{Fill in the Description}}
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/help/Enable-AzureRmCdnCustomDomain.md]: @prakharsharma10 please add a description for this cmdlet

File: src/ResourceManager/Cdn/Commands.Cdn/help/Get-AzureRMCdnProfile.md [hunk 14]
```diff
@@ -80,9 +80,6 @@ This cmdlet supports the common parameters: -Debug, -ErrorAction, -ErrorVariable
 
 ## INPUTS
 
-### None
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/help/Get-AzureRMCdnProfile.md]: @prakharsharma10 please revert the change made here

File: src/ResourceManager/Cdn/Commands.Cdn/help/Get-AzureRmCdnEndpointNameAvailability.md [hunk 15]
```diff
@@ -65,14 +65,10 @@ This cmdlet supports the common parameters: -Debug, -ErrorAction, -ErrorVariable
 
 ## INPUTS
 
-### None
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/help/Get-AzureRmCdnEndpointNameAvailability.md]: @prakharsharma10 please revert the change made here

File: src/ResourceManager/Cdn/Commands.Cdn/help/Get-AzureRmCdnProfileSupportedOptimizationType.md [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+```
+
+### ByObjectParameterSet
+```
+Get-AzureRmCdnProfileSupportedOptimizationType -CdnProfile <PSProfile>
+ [-DefaultProfile <IAzureContextContainer>] [<CommonParameters>]
+```
+
+## DESCRIPTION
+{{Fill in the Description}}
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/help/Get-AzureRmCdnProfileSupportedOptimizationType.md]: @prakharsharma10 please add a description for this cmdlet

File: src/ResourceManager/Cdn/Commands.Cdn/help/New-AzureRmCdnProfile.md [hunk 17]
```diff
@@ -157,9 +157,6 @@ This cmdlet supports the common parameters: -Debug, -ErrorAction, -ErrorVariable
 
 ## INPUTS
 
-### None
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/help/New-AzureRmCdnProfile.md]: @prakharsharma10 please revert the change made here

File: src/ResourceManager/Cdn/Commands.Cdn/packages.config [hunk 18]
```diff
@@ -1,4 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-	<package id="Microsoft.Azure.Management.Cdn" version="3.0.0-preview" targetFramework="net45" />
-</packages>
+  <package id="Microsoft.Azure.Management.Cdn" version="4.0.1-preview" targetFramework="net452" />
+  <package id="Microsoft.Rest.ClientRuntime" version="2.3.10" targetFramework="net452" />
```
Comment (cormacpayne) [inline src/ResourceManager/Cdn/Commands.Cdn/packages.config]: @prakharsharma10 similar to the `.csproj` file, remove the references added in this file (`Microsoft.Rest.ClientRuntime.*` and `Newtonsoft.Json`) since they are referenced in a separate `packages.config`

File: src/ResourceManager/Cdn/Commands.Cdn/help/Get-AzureRMCdnProfile.md (hunk 14 above)
Comment (prakharsharma10) [reply src/ResourceManager/Cdn/Commands.Cdn/help/Get-AzureRMCdnProfile.md]: This file is auto-generated using the platyPS module

File: src/ResourceManager/Cdn/Commands.Cdn/help/Get-AzureRmCdnEndpointNameAvailability.md (hunk 15 above)
Comment (prakharsharma10) [reply src/ResourceManager/Cdn/Commands.Cdn/help/Get-AzureRmCdnEndpointNameAvailability.md]: This file is auto-generated using the platyPS module

File: src/ResourceManager/Cdn/Commands.Cdn/help/New-AzureRmCdnProfile.md (hunk 17 above)
Comment (prakharsharma10) [reply src/ResourceManager/Cdn/Commands.Cdn/help/New-AzureRmCdnProfile.md]: This file is auto-generated using the platyPS module

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/ConfirmAzureRmCdnEndpointProbeUrl.cs (hunk 7 above)
Comment (prakharsharma10) [reply src/ResourceManager/Cdn/Commands.Cdn/Endpoint/ConfirmAzureRmCdnEndpointProbeUrl.cs]: This command validates the probe URL. We have the implementations in the REST API and SDKs which use the name "validate". Have used a substitute for that as listed here https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.verbslifecycle.confirm?redirectedfrom=MSDN&view=powershellsdk-1.1.0

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/NewAzureRmCdnEndpoint.cs (hunk 9 above)
Comment (prakharsharma10) [reply src/ResourceManager/Cdn/Commands.Cdn/Endpoint/NewAzureRmCdnEndpoint.cs]: Delivery Policy is not a resource, it's a property. The users can create this object as shown in the EndpointTests.ps1, the same way they create geofilters

File: src/ResourceManager/Cdn/Commands.Cdn/Profile/GetAzureRmCdnProfileSupportedOptimizationType.cs (hunk 10 above)
Comment (prakharsharma10) [reply src/ResourceManager/Cdn/Commands.Cdn/Profile/GetAzureRmCdnProfileSupportedOptimizationType.cs]: Thanks for the suggestion :) . It's a design decision and we have gone through this in the past here https://github.com/Azure/azure-rest-api-specs/pull/1523/files/ef75b2c5a9a237c413122f141c4b6c3722ac103d#diff-6bb303ad17bff3c22d65d5ab3ee7b8cf . Since the supported optimization types can change without any update from the user side, we want to keep it separate from the profile entity. For example, if we return the profile along with supported optimization types, the profile as a resource will keep changing from time to time as we change the types for the SKUs.

**Comment by prakharsharma10** [comment]
@cormacpayne Have addressed all the comments. The build is failing because of some unrelated scenario tests. Please take look. Thanks!

**Comment by prakharsharma10** [comment]
@cormacpayne I pulled the changed from the preview branch for the updated clientruntime, resolved the conflicts and re-recorded the tests but the tests are failing.

This is because the long-running operations (LRO) are handled differently in the updated client runtime which our CDN service does not support. I'm not sure why it has been updated for our service. Regarding the conflicts, the recorded files that are already checked in the upstream repo have failed recordings. These should not have been checked in.

Please let us know how we can use the existing clientruntime assemblies and also review the addressed comments at the earliest. Thanks!

**Comment by cormacpayne** [comment]
Closing in favor of https://github.com/Azure/azure-powershell/pull/5855
